Bottom: 8f25c7660f0e4b9821a48ab9e1ab569ee05a2aa2
Top:    4e08c13b08a976b4f69198abe6b84e7345719717
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-30 18:00:57 +0100

Refresh of puller-c-re-write-to-use-hash-

---

diff --git a/src/puller.c b/src/puller.c
index 6ee8ded..a2c0b35 100644
--- a/src/puller.c
+++ b/src/puller.c
@@ -634,7 +634,6 @@ typedef struct Extra {
 static Extra multi_next;
 static GHashTable *lines;
 static GHashTable *arcs;
-static int nlines, narcs;
 static int did_something;
 static int current_is_component, current_is_solder;
 
@@ -1040,16 +1039,34 @@ null_multi_next_ends (AnyObjectType *ptr, Extra *extra, void *userdata)
     extra->end.next = NULL;
 }
 
+static Extra *
+new_line_extra (LineType *line)
+{
+  Extra *extra = g_slice_new0 (Extra);
+  g_hash_table_insert (lines, line, extra);
+  extra->parent.line = line;
+  return extra;
+}
+
+static Extra *
+new_arc_extra (ArcType *arc)
+{
+  Extra *extra = g_slice_new0 (Extra);
+  g_hash_table_insert (arcs, arc, extra);
+  extra->parent.arc = arc;
+  return extra;
+}
+
 static void
 find_pairs ()
 {
   ARC_LOOP (CURRENT); {
-    Extra *e = ARC2EXTRA (arc);
+    Extra *e = new_arc_extra (arc);
     fix_arc_extra (arc, e);
   } END_LOOP;
 
   LINE_LOOP (CURRENT); {
-    Extra *e = LINE2EXTRA (line);
+    Extra *e = new_line_extra (line);
     if (line->Point1.X >= 0)
       {
 	find_pairs_1 (line, & e->start.next, line->Point1.X, line->Point1.Y);
@@ -1816,11 +1833,7 @@ create_line (LineTypePtr sample, int x1, int y1, int x2, int y2)
 					   sample->Thickness, sample->Clearance, sample->Flags);
   AddObjectToCreateUndoList (LINE_TYPE, CURRENT, line, line);
 
-  nlines = CURRENT->LineN;
-
-  e = g_slice_new0 (Extra);
-  g_hash_table_insert (lines, e, line);
-
+  e = new_line_extra (line);
 #if TRACE1
   printf(" - line extra is %p\n", e);
 #endif
@@ -1852,11 +1865,7 @@ create_arc (LineTypePtr sample, int x, int y, int r, int sa, int da)
   if (!arc)
     longjmp (abort_buf, 1);
 
-  narcs = CURRENT->ArcN;
-
-  e = g_slice_new0 (Extra);
-  g_hash_table_insert (arcs, e, arc);
-
+  e = new_arc_extra (arc);
 #if TRACE1
   printf(" - arc extra is %p\n", e);
 #endif
